File: python/LinkedList/Scaler/reverse_kth_group.py
class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next


def reverse_kth_group(head, k):
    count = 0
    temp = head

    while temp and count < k:
        temp = temp.next
        count += 1

    if count < k:
        return head

    prev = None
    current = head

    for _ in range(k):
        next_node = current.next
        current.next = prev
        prev = current
        current = next_node

    # Recursively reverse the next group
    head.next = reverse_kth_group(current, k)
    return prev

# Function to print the linked list


def print_list(head):
    current = head
    while current:
        print(current.val, end=" ")
        current = current.next
    print()


# Example usage
# Create a linked list: 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 7 -> 8 -> None
root = Node(1)
root.next = Node(2)
root.next.next = Node(3)
root.next.next.next = Node(4)
root.next.next.next.next = Node(5)
root.next.next.next.next.next = Node(6)
root.next.next.next.next.next.next = Node(7)
root.next.next.next.next.next.next.next = Node(8)

print("Original linked list:")
print_list(root)

# Reverse the k group nodes
k = 3
root = reverse_kth_group(root, k)

print("After reversing the first", k, "nodes:")
print_list(root)

File: JavaScript/LinkedList/Scaler/reverse_kth_group.js
class Node {
  constructor(val, next = null) {
    this.val = val;
    this.next = next;
  }
}

function reverse_k_group_nodes(head, k) {
  let temp = head;
  let count = 0;

  while (temp && count < k) {
    temp = temp.next;
    count += 1;
  }

  if (count < k) {
    return head;
  }

  let curr = head;
  let prev = null;

  for (let i = 0; i < k; i++) {
    let nextNode = curr.next;
    curr.next = prev;
    prev = curr;
    curr = nextNode;
  }

  head.next = reverse_k_group_nodes(curr, k);

  return prev;
}

// # Function to print the linked list
function printList(head) {
  let current = head;
  while (current) {
    process.stdout.write(current.val + " -> "); // Equivalent to Python's end=" "
    current = current.next;
  }
  console.log(); // Print a newline character to move to the next line
}

//  Example usage
//   Create a linked list: 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 7 -> 8 -> None
let root = new Node(1);
root.next = new Node(2);
root.next.next = new Node(3);
root.next.next.next = new Node(4);
root.next.next.next.next = new Node(5);
root.next.next.next.next.next = new Node(6);
root.next.next.next.next.next.next = new Node(7);
root.next.next.next.next.next.next.next = new Node(8);

console.log("Original linked list:");
printList(root);

const k = 3;
// # Reverse Linked List
const rootNode = reverse_k_group_nodes(root, k);

console.log("After reversing " + k + " group nodes:");
printList(rootNode);
